### PR TITLE
Use info config for stats computation

### PR DESCRIPTION
**Pull request by VianneyRuhlmann**
# What does this PR do?
Fetch the info endpoint from the trace exporter to use configuration from the agent for client-side stats. The trace exporter will also check that the agent support client side stats before enabling it.

# Motivation

The agent info contains essential configuration for peer tags aggregation and span kinds eligibility to stats.

# Additional Notes
## Mutability concerns
Since we check the agent info on each call to `send` to update `StatsComputationStatus` I used an ArcSwap (which we already use for the info struct) to wrap the stats computation status. This allows sending traces without holding a mutable reference. ArcSwap is optimized for read-mostly scenarios which is the case here. This should reduce the impact of having to lock the status on every call to `send` to check if stats should be computed.

If you have any thoughts about this solution please let me know, fyi here's the other solutions I considered
### Other solutions considered
- Let `send` take a &mut and use a Mutex on the whole `TraceExporter` this is not viable as send block on the http calls which means holding the lock for a long time
- Using a `Mutex` or a `RwLock` a bit expensive considering it would be on the critical path of send

# How to test the change?

Describe here in detail how the change can be validated.

### CODE REVIEW

File: data-pipeline/src/span_concentrator/mod.rs
Comment (ajgajg1134) [inline data-pipeline/src/span_concentrator/mod.rs:71]:
```suggestion
    /// span.kind fields eligible for stats computation
```

File: data-pipeline/src/trace_exporter.rs
Comment (ajgajg1134) [inline data-pipeline/src/trace_exporter.rs:300]: Should we add a note here that this won't work if stats computation is disabled at the client level?

File: data-pipeline/src/trace_exporter.rs
Comment (ajgajg1134) [inline data-pipeline/src/trace_exporter.rs:363]: do you want to keep this `dbg!` in?

File: data-pipeline/src/trace_exporter.rs
Comment (ajgajg1134) [inline data-pipeline/src/trace_exporter.rs:260]:
```suggestion
                       Arc::<StatsComputationStatus>::into_inner(
```

File: data-pipeline/src/trace_exporter.rs
Comment (ajgajg1134) [inline data-pipeline/src/trace_exporter.rs:281]:
```suggestion
                   Arc::<StatsComputationStatus>::into_inner(
```

**Review by ajgajg1134 — COMMENTED** [review]
Overall this makes sense to me, just a few small clean up comments (probably want another rustacean to take a peek at this before merging though).

Some of the syntax here is a bit silly looking, but I think it's just kinda "required" from what I can tell (like `&**self.client_side_stats.load()`). Maybe this is just me still getting used to reading rust syntax again :P

File: data-pipeline/src/trace_exporter.rs
Comment (ekump) [inline data-pipeline/src/trace_exporter.rs:1213]: since this is in a loop, would it make sense to lower the duration?

File: data-pipeline/src/trace_exporter.rs
Comment (ekump) [inline data-pipeline/src/trace_exporter.rs:32]: I wonder if we're not properly encapsulating the stats computation logic? It's not obvious to me why all of these changes should live in the trace exporter. These consts , the `StatsComputationStatus` enum, and the stats specific start/shutdown logic are all examples.

File: data-pipeline/src/trace_exporter.rs
Comment (ekump) [inline data-pipeline/src/trace_exporter.rs:362]: Is it possible that this (and other large functions as part of this PR) be broken up into smaller ones to make them easier to read and maintain?

File: data-pipeline/src/trace_exporter.rs
Comment (VianneyRuhlmann) [reply data-pipeline/src/trace_exporter.rs:1213]: I tried a lower duration with the info fetcher but this sometimes failed probably due to a race condition where the loop ended because the server received the request but the response had not yet been received by the client.

File: data-pipeline/src/trace_exporter.rs
Comment (bantonsson) [inline data-pipeline/src/trace_exporter.rs:412]: So why does this code differ from the code above, where the span kinds use this default, but here it's added to the peer tags 🤷🏼‍♂️

**Review by bantonsson — APPROVED** [review]
LGTM. Only one weird thing with the peer tag default fallback.

File: data-pipeline/src/trace_exporter.rs
Comment (VianneyRuhlmann) [reply data-pipeline/src/trace_exporter.rs:32]: I agree, yet this would probably require a larger refactor. I'll try to improve it without refactoring too much.

**Review by knusbaum — APPROVED** [review]

**Review by ekump — APPROVED** [review]
LGTM. As discussed IRL we're going to consider a bigger refactor of the trace exporter and don't want to block on that right now.

File: data-pipeline/src/trace_exporter.rs
Comment (VianneyRuhlmann) [reply data-pipeline/src/trace_exporter.rs:298]: I think it's fine if the the client didn't enable stats computation there's no point for the exporter to start it and since it is private I don't think there's a risk of misusing it

File: data-pipeline/src/trace_exporter.rs
Comment (VianneyRuhlmann) [reply data-pipeline/src/trace_exporter.rs:409]: That's just me going crazy, I fixed it.

**Pull request merged**
